crates/wasmtime/src/component/types.rs

```diff
@@ -0,0 +1,1025 @@
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub struct UnionIndex(TypeUnionIndex);
+
+impl Handle<UnionIndex> {
+    pub fn types(&self) -> impl ExactSizeIterator<Item = Type> + '_ {
+        self.types[self.index.0]
+            .types
+            .iter()
+            .map(|ty| Type::from(ty, &self.types))
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub struct ExpectedIndex(TypeExpectedIndex);
+
+impl Handle<ExpectedIndex> {
+    pub fn ok(&self) -> Type {
+        Type::from(&self.types[self.index.0].ok, &self.types)
+    }
+
+    pub fn err(&self) -> Type {
+        Type::from(&self.types[self.index.0].err, &self.types)
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub struct FlagsIndex(TypeFlagsIndex);
+
+impl Handle<FlagsIndex> {
+    pub fn names(&self) -> impl ExactSizeIterator<Item = &str> {
+        self.types[self.index.0]
+            .names
+            .iter()
+            .map(|name| name.deref())
+    }
+}
+
+/// Represents the size and alignment requirements of the heap-serialized form of a type
+pub(crate) struct SizeAndAlignment {
+    pub(crate) size: usize,
+    pub(crate) alignment: u32,
+}
+
+/// Represents a component model interface type
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum Type {
+    /// Unit
+    Unit,
+    /// Boolean
+    Bool,
+    /// Signed 8-bit integer
+    S8,
+    /// Unsigned 8-bit integer
+    U8,
+    /// Signed 16-bit integer
+    S16,
+    /// Unsigned 16-bit integer
+    U16,
+    /// Signed 32-bit integer
+    S32,
+    /// Unsigned 32-bit integer
+    U32,
+    /// Signed 64-bit integer
+    S64,
+    /// Unsigned 64-bit integer
+    U64,
+    /// 64-bit floating point value
+    Float32,
+    /// 64-bit floating point value
+    Float64,
+    /// 32-bit character
+    Char,
+    /// Character string
+    String,
+    /// List of values
+    List(Handle<TypeIndex>),
+    /// Record
+    Record(Handle<RecordIndex>),
+    /// Tuple
+    Tuple(Handle<TupleIndex>),
+    /// Variant
+    Variant(Handle<VariantIndex>),
+    /// Enum
+    Enum(Handle<EnumIndex>),
+    /// Union
+    Union(Handle<UnionIndex>),
+    /// Option
+    Option(Handle<TypeIndex>),
+    /// Expected
+    Expected(Handle<ExpectedIndex>),
+    /// Bit flags
+    Flags(Handle<FlagsIndex>),
+}
+
+impl Type {
+    /// Instantiate this type (which must be a `Type::List`) with the specified `values`.
+    pub fn new_list(&self, values: Box<[Val]>) -> Result<Val> {
+        if let Type::List(handle) = self {
+            let ty = handle.ty();
+            for (index, value) in values.iter().enumerate() {
+                ty.check(value)
+                    .with_context(|| format!("type mismatch for element {index} of list"))?;
+            }
+
+            Ok(Val::List(List {
+                ty: handle.clone(),
+                values,
+            }))
+        } else {
+            Err(anyhow!("cannot make list from {} type", self.desc()))
+        }
+    }
```

Comment:
I think these methods may make more sense on `List` (in the values module) along the lines of:


```rust
impl List {
    pub fn new(elements: Box<[Val]>, ty: types::List) -> Result<List> {
        // ...
    }
}
```

(using the different structure of types I mentioned above).

Then there could be `impl From<List> for Val` to convert to a `Val` as necessary.

Reply:
That's what I originally considered, but I ended up using these methods on `Type` instead in the interest of ergonomics.  You can see them used in the dynamic.rs tests, along with `Type::nested`.  This avoids forcing the programmer to write a bunch of `if let Type::List(ty) = ty { List::new(elements, ty) } else { unreachable!() }`.

This is the ergonomic challenge I was referring to earlier in the discussion about whether `Val`s need to be associated with a specific `Type`.  I'm certainly willing to reconsider; I just want to understand how you expect this API to be used vs how I'm using it in the dynamic.rs tests.  Do you not expect real-world code to need a bunch of `if let ... { ... } else { unreachable!() }` when building `Val`s from `Type`s in the scheme you outlined?

Reply:
Ah ok thanks for pointing that out! I think this may be a case where I'm used to slightly more verbose Rust-based idioms perhaps? For example the `Type::new_list` API as-is can return two kinds of errors (not a list type or element of the wrong type) where what I would otherwise expect is to split those two error cases across two functions. Instead of:

```rust
    let input = ty.new_list(Box::new([
        Val::U32(32343),
        Val::U32(79023439),
        Val::U32(2084037802),
    ]))?;
```

I'd expect:

```rust
    let input = ty.unwrap_list().new(Box::new([
        Val::U32(32343),
        Val::U32(79023439),
        Val::U32(2084037802),
    ]))?;
```

(or similar)

I don't mean to say we should go all the way to writing a bunch of wordy `if let` but I think helper methods can help scope precisely what errors are coming out of each function. In this case `unwrap_list` would be a clear indication of "I expect this type to be a list" and the `new` error would be clearly "I expect each element to be of this type".